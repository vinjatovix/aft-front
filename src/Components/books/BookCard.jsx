import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";

import { ListItem } from "../common/ListItem";
import { NavigateButton } from "../ui/buttons/NavigateButton";
import { AdminButtons } from "../ui/admin/AdminButtons";

const DEFAULT_COVER_BOOK =
  "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";

export const BookCard = ({ data, isEditor, actions, token }) => (
  <div className="flex-wrapper column">
    <ul
      data-testid="book-detail"
      className="book"
      onClick={() => {
        actions.detail();
      }}
    >
      {<img src={data?.img || DEFAULT_COVER_BOOK} alt={data.name} />}
      <ListItem name="Obra" children={data.name} />
      <ListItem name="Autor" className="small" children={data.author} />
    </ul>

    <NavLink to={`/book/${data._id}/characters`}>
      <NavigateButton text="ðŸŽ­ Personajes" />
    </NavLink>

    <NavLink to={`/book/${data._id}/scenes`}>
      <NavigateButton text="ðŸŽ¬ Escenas" />
    </NavLink>

    {isEditor && <AdminButtons token={token} item={data} actions={actions} />}
  </div>
);

BookCard.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    img: PropTypes.string,
    description: PropTypes.string.isRequired,
  }).isRequired,
  isEditor: PropTypes.bool,
  actions: PropTypes.shape({
    setDataDetail: PropTypes.func.isRequired,
    add: PropTypes.func.isRequired,
    close: PropTypes.func.isRequired,
    delete: PropTypes.func.isRequired,
    edit: PropTypes.func.isRequired,
    detail: PropTypes.func.isRequired,
    refresh: PropTypes.func.isRequired,
  }).isRequired,
  token: PropTypes.string.isRequired,
};

BookCard.defaultProps = {
  isEditor: false,
};
