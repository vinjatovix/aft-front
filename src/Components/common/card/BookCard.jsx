import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import { ListItem } from "../ListItem";

export const BookCard = ({ data, isAdmin, actions, token }) => {
  const defaultCoverBook =
    "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";
  return (
    <div className="flex-wrapper column">
      <ul
        className="book"
        onClick={() => {
          actions.detail();
        }}
      >
        {<img src={data.img ? data.img : defaultCoverBook} alt={data.name} />}
        <ListItem name="Obra" children={data.name} />
        <ListItem name="Autor" className="small" children={data.author} />
      </ul>
      <NavLink to={`/characters/book/${data._id}`}>
        <button data-testid="char-button" className="bg-blue">
          🎭 Personajes
        </button>
      </NavLink>
      {isAdmin && (
        <div className="control-panel-mini">
          <button
            className="bg-yellow"
            data-testid="edit-button"
            onClick={() => {
              actions.edit();
            }}
          >
            📝 editar
          </button>
          <button
            data-testid="delete-button"
            className="bg-red"
            onClick={() => {
              actions.delete(token, data._id);
            }}
          >
            🗑️ eliminar
          </button>
        </div>
      )}
    </div>
  );
};

BookCard.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    img: PropTypes.string,
  }).isRequired,
  isAdmin: PropTypes.bool,
  actions: PropTypes.object.isRequired,
  token: PropTypes.string.isRequired,
};

BookCard.defaultProps = {
  isAdmin: false,
};
