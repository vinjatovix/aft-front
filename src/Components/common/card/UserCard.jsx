import React from "react";
import PropTypes from "prop-types";

import { ListItem } from "../ListItem";
import { UserRole } from "../../ui/UserRole";

export const UserCard = ({ data }) => {
  return (
    <ul className="user-list">
      <ListItem name="Nombre" children={data.username} />
      <ListItem name="Role" children={UserRole(data.roles)} />
    </ul>
  );
};

UserCard.propTypes = {
  data: PropTypes.shape({
    username: PropTypes.string.isRequired,
    roles: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
};
