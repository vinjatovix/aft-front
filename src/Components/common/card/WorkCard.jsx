import React from "react";
import PropTypes from "prop-types";
import { ListItem } from "../ListItem";

export const WorkCard = ({ data }) => {
  return (
    <ul className="work-list">
      <ListItem
        name="Intérprete"
        children={<p>by: {data.metadata.createdBy}</p>}
        liKey={data.metadata.createdBy}
      />
      <ListItem
        name="Obra"
        children={<p>Obra: {data.character.book.name}</p>}
        liKey={data.character.book._id}
      />
      <ListItem
        name="Personaje"
        children={<p>Personaje: {data.character.name}</p>}
        liKey={data.character._id}
      />
      <ListItem
        name="Escena"
        children={<p>Escena: {data.scene.name}</p>}
        liKey={data.scene._id}
      />
      <ListItem
        name="Descripción"
        children={<p>Descripción: {data.description}</p>}
      />
    </ul>
  );
};

WorkCard.propTypes = {
  data: PropTypes.object.isRequired,
};
