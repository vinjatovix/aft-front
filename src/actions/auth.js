import { ACTION_TYPES } from "../types/types";

export const loginRequestAction = () => ({
  type: ACTION_TYPES.AUTH_LOGIN_REQUEST,
});

export const loginSuccessAction = (auth) => ({
  type: ACTION_TYPES.AUTH_LOGIN_SUCCESS,
  payload: auth,
});

export const loginFailureAction = (error) => ({
  type: ACTION_TYPES.AUTH_LOGIN_FAILURE,
  payload: error,
});

export const logoutAction = () => ({
  type: ACTION_TYPES.AUTH_LOGOUT,
});
