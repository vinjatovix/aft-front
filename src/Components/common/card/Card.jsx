import React from "react";
import PropTypes from "prop-types";
import { UserCard } from "./UserCard";

const Card = ({ data, type, actions, isAdmin, token }) => {
  return (
    <div className="card" onClick={() => actions.setDataDetail(data)}>
      {type === "user" && <UserCard data={data} />}
    </div>
  );
};

Card.propTypes = {
  type: PropTypes.string.isRequired,
  data: PropTypes.object.isRequired,
  actions: PropTypes.object,
  isAdmin: PropTypes.bool,
  token: PropTypes.string,
};

Card.defaultProps = {
  isAdmin: false,
};

export default Card;
