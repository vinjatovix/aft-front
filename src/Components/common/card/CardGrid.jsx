import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { modalActionsType, modalsStateType } from "../../../PropTypes";
import { api } from "../../../http/api";

import { Card } from "./Card";
import { ContentHeader } from "../content/ContentHeader";

import { useFetch } from "../../../hooks/useFetch";

export const CardGrid = ({
  actions,
  auth: { token },
  isEditor,
  modals,
  type,
}) => {
  const [{ data, loading, error }, reFetch] = useFetch(
    api[type].getAll.path,
    api[type].getAll.method,
    { token }
  );

  const title = type === "work" ? "Trabajos:" : "Obras:";
  const count = data ? data.length : 0;

  useEffect(() => {
    token && reFetch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [modals.lastRefresh, token]);

  return (
    <>
      <ContentHeader count={count} title={title} />
      <div className="card-grid">
        {data &&
          data.map((item) => (
            <Card
              actions={actions}
              data={item}
              isEditor={isEditor}
              key={item._id}
              token={token}
              type={type}
            />
          ))}
        {error && <div>{error}</div>}
        {loading && <div>Cargando...</div>}
      </div>
    </>
  );
};

CardGrid.propTypes = {
  actions: modalActionsType,
  auth: PropTypes.shape({
    token: PropTypes.string.isRequired,
  }),
  isEditor: PropTypes.bool,
  modals: modalsStateType,
  type: PropTypes.string.isRequired,
};

CardGrid.defaultProps = {
  isEditor: false,
};
