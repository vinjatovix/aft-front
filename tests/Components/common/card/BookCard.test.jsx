import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import {
  BrowserRouter as Router,
  MemoryRouter,
  Route,
  Routes,
} from "react-router-dom";
import { BookCard } from "../../../../src/Components/common/card/BookCard";
import { UserContext } from "../../../../src/contexts/UserContext";
import actions from "../../../fixtures/actions";
import getUserContext from "../../../fixtures/getUserContext";
import { arrayElement } from "../../../fixtures/random";

describe("BookCard Component", () => {
  const data = {
    _id: "5e9f8f8f8f8f8f8f8f8f8f8",
    name: "Obra",
    author: "Autor",
    img: "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg",
    description: "Descripción",
  };

  it("should match snapshot", async () => {
    const { container } = render(
      <Router>
        <BookCard data={data} isAdmin={true} actions={actions} token="token" />
      </Router>
    );

    expect(container).toMatchSnapshot();
  });

  it("should show author inside li class small", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={false} actions={actions} token="token" />
      </Router>
    );
    const li = screen.getByTestId(data.author);

    expect(li.classList.contains("small")).toBeTruthy();
    expect(li.innerHTML).toContain(data.author);
  });

  it("should show only characters and scenes buttons (non admin card)", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={false} actions={actions} token="token" />
      </Router>
    );
    const buttons = screen.getAllByRole("button");

    expect(buttons.length).toBe(2);
  });

  it("should navigate to characters when character button is clicked", async () => {
    render(
      <UserContext.Provider
        value={{
          auth: getUserContext(
            arrayElement(["aft-admin", "aft-user", "aft-editor"])
          ),
        }}
      >
        <MemoryRouter initialEntries={["/books"]}>
          <Routes>
            <Route
              path="/books"
              element={
                <BookCard
                  data={data}
                  isAdmin={true}
                  actions={actions}
                  token="token"
                />
              }
            />
            <Route
              path="/characters/book/:bookId"
              element={<h1>CharactersScreen</h1>}
            />
          </Routes>
        </MemoryRouter>
      </UserContext.Provider>
    );

    const button = screen.getAllByTestId("navigate-button");
    fireEvent.click(button[0]);

    expect(screen.getByText("CharactersScreen")).toBeTruthy();
  });

  it("should show characters, scenes, edit and delete buttons (admin card)", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={true} actions={actions} token="token" />
      </Router>
    );
    const buttons = screen.getAllByTestId("action-button");
    const editButton = buttons[0];
    const deleteButton = buttons[1];
    const navigateButtons = screen.getAllByTestId("navigate-button");

    expect(buttons.length).toBe(2);
    expect(editButton).toBeTruthy();
    expect(editButton.innerHTML).toBe("📝 Editar");
    expect(deleteButton).toBeTruthy();
    expect(deleteButton.innerHTML).toBe("🗑️ Eliminar");
    expect(navigateButtons[0]).toBeTruthy();
    expect(navigateButtons[0].innerHTML).toBe("🎭 Personajes");
    expect(navigateButtons[1]).toBeTruthy();
    expect(navigateButtons[1].innerHTML).toBe("🎬 Escenas");
  });

  it("should call actions.edit when button is clicked", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={true} actions={actions} token="token" />
      </Router>
    );
    const buttons = screen.getAllByTestId("action-button");

    fireEvent.click(buttons[0]);

    expect(actions.edit).toHaveBeenCalled();
  });

  it("should call actions.delete when button is clicked", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={true} actions={actions} token="token" />
      </Router>
    );

    const buttons = screen.getAllByTestId("action-button");
    fireEvent.click(buttons[1]);

    expect(actions.delete).toHaveBeenCalledWith(
      "token",
      "5e9f8f8f8f8f8f8f8f8f8f8"
    );
  });

  it("should call actions.detail when book ul is clicked", async () => {
    render(
      <Router>
        <BookCard data={data} isAdmin={true} actions={actions} token="token" />
      </Router>
    );

    const ul = screen.getByTestId("book-detail");
    fireEvent.click(ul);

    expect(actions.detail).toHaveBeenCalled();
  });
});
