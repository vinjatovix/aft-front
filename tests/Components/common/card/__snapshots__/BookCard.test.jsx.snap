// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard Component should match snapshot 1`] = `
<div>
  <div
    class="flex-wrapper column"
  >
    <ul
      class="book"
      data-testid="book-detail"
    >
      <img
        alt="Obra"
        src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
      />
      <li
        class=""
        data-testid="Obra"
      >
        Obra
      </li>
      <li
        class="small"
        data-testid="Autor"
      >
        Autor
      </li>
    </ul>
    <a
      class=""
      href="/book/5e9f8f8f8f8f8f8f8f8f8f8/characters"
    >
      <button
        class="bg-blue"
        data-testid="navigate-button"
      >
        🎭 Personajes
      </button>
    </a>
    <a
      class=""
      href="/book/5e9f8f8f8f8f8f8f8f8f8f8/scenes"
    >
      <button
        class="bg-blue"
        data-testid="navigate-button"
      >
        🎬 Escenas
      </button>
    </a>
    <div
      class="control-panel-mini "
    >
      <button
        class="btn-big bg-yellow"
        data-testid="action-button"
      >
        📝 Editar
      </button>
      <button
        class="btn-big bg-red"
        data-testid="action-button"
      >
        🗑️ Eliminar
      </button>
    </div>
  </div>
</div>
`;
